service/worker/deployment/deployment_reachability.go

```diff
@@ -0,0 +1,183 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2024 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deployment
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	deploymentpb "go.temporal.io/api/deployment/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/server/common/cache"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/common/searchattribute"
+	"go.temporal.io/server/common/worker_versioning"
+)
+
+const (
+	reachabilityCacheOpenWFsTTL   = time.Minute
+	reachabilityCacheClosedWFsTTL = 10 * time.Minute
+	reachabilityCacheMaxSize      = 10000
+)
+
+// We update the reachability search attribute on workflow task completion and when there is
+// an UpdateWorkflowExecutionOptions request that changes the workflow's effective behavior
+// or deployment. We do not update the reachability search attribute if the search attribute
+// has not changed. We also do not update the reachability search attribute if a workflow has
+// started transitioning to a new deployment, because only on task completion do we know whether
+// the transition has succeeded or failed.
+func getDeploymentReachability(
+	ctx context.Context,
+	nsId, nsName, seriesName, buildID, currentBuildID string,
+	currentBuildIDValidTime time.Time,
+	rc reachabilityCache,
+) (enumspb.DeploymentReachability, time.Time, error) {
+	// 1a. Reachable by new unpinned workflows
+	if buildID == currentBuildID { // add if buildID is ramping, once we have ramp
+		return enumspb.DEPLOYMENT_REACHABILITY_REACHABLE, currentBuildIDValidTime, nil
```

Comment:
do we need a TODO for "still return reachable if the deployment just became not current"?

Reply:
do you mean, because there are still outstanding tasks on that deployment?

Reply:
some new wfs might have started at t0 on the deployment when it was current. Not is t0.001 and does visibility records are still not indexed. here we'd say unreachable because the deployment is not current and reachability count is 0.

Reply:
ah yes. I was trying to remember how to use the `visibilityGracePeriod` last night and couldnt remember so I left it out to keep things simple. but I will put a todo